unified_planning/model/scheduling/scheduling_problem.py

```diff
@@ -0,0 +1,223 @@
+        s.append("]\n\n")
+        if len(self.user_types) > 0:
+            s.append("objects = [\n")
+            for ty in self.user_types:
+                s.append(f"  {str(ty)}: {str(list(self.objects(ty)))}\n")
+            s.append("]\n\n")
+        s.append("initial fluents default = [\n")
+        for f in self._fluents:
+            if f in self._fluents_defaults:
+                v = self._fluents_defaults[f]
+                s.append(f"  {str(f)} := {str(v)}\n")
+        s.append("]\n\n")
+        s.append("initial values = [\n")
+        for k, v in self.explicit_initial_values.items():
+            s.append(f"  {str(k)} := {str(v)}\n")
+        s.append("]\n\n")
+        if len(self.quality_metrics) > 0:
+            s.append("quality metrics = [\n")
+            for qm in self.quality_metrics:
+                s.append(f"  {str(qm)}\n")
+            s.append("]\n")
+        s.append("\nBASE")
+        s.append(str(self._base))
+
+        s.append("\n\nActivities:\n  ")
+        for act in self._activities:
+            s.append(str(act))
+            s.append("\n  ")
+
+        return "".join(s)
+
+    def __eq__(self, oth: object) -> bool:
+        if not (isinstance(oth, SchedulingProblem)) or self._env != oth._env:
+            return False
+        if self.kind != oth.kind or self._name != oth._name:
+            return False
+
+        if not UserTypesSetMixin.__eq__(self, oth):
+            return False
+        if not ObjectsSetMixin.__eq__(self, oth):
+            return False
+        if not FluentsSetMixin.__eq__(self, oth):
+            return False
+        if not InitialStateMixin.__eq__(self, oth):
+            return False
+        if not MetricsMixin.__eq__(self, oth):
+            return False
+
+        if self._base != oth._base:
+            return False
+        if set(self._activities) != set(oth._activities):
+            return False
+        return True
+
+    def __hash__(self) -> int:
+        res = hash(self.kind) + hash(self._name)
+
+        res += UserTypesSetMixin.__hash__(self)
+        res += ObjectsSetMixin.__hash__(self)
+        res += FluentsSetMixin.__hash__(self)
+        res += InitialStateMixin.__hash__(self)
+        res += MetricsMixin.__hash__(self)
+
+        res += hash(self._base)
+        res += sum(map(hash, self._activities))
+        return res
+
+    def clone(self):
+        new_p = SchedulingProblem(self._name, self._env)
+        UserTypesSetMixin._clone_to(self, new_p)
+        ObjectsSetMixin._clone_to(self, new_p)
+        FluentsSetMixin._clone_to(self, new_p)
+        InitialStateMixin._clone_to(self, new_p)
+        MetricsMixin._clone_to(self, new_p, new_actions=None)
+
+        new_p._base = self._base.clone()
+        new_p._activities = [a.clone() for a in self._activities]
+        return new_p
+
+    def has_name(self, name: str) -> bool:
+        return name in self._base._parameters
+
+    def add_variable(self, name: str, tpe: Type) -> Parameter:
```

Comment:
It is not clear to me what is the semantic of a variable here.

Reply:
A scheduling problem is essentially a CSP, for which the UP provides has convenient abstraction.

A variable is simple  a decision variable. More precisely, it can be seen as an existentially quantified variable whose scope is the entire problem.

Reply:
Can you add a comment explaining this?